File: api-servidor/src/infrastructure/database/models/company-model.ts
import { DataTypes, Model, Optional } from 'sequelize';
import { sequelize } from '../config/connection';

interface CompanyAttributes {
  id: string;
  userId: string;
  tradeName: string;
  legalName: string;
  cnpj: string | null;
  phone: string;
  address: string;
  city: string;
  state: string;
  description: string | null;
  logoUrl: string | null;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
  deletedAt: Date | null;
}

type CompanyCreationAttributes = Optional<CompanyAttributes, 'id' | 'createdAt' | 'updatedAt' | 'deletedAt'>;

class CompanyModel
  extends Model<CompanyAttributes, CompanyCreationAttributes>
  implements CompanyAttributes
{
  public id!: string;
  public userId!: string;
  public tradeName!: string;
  public legalName!: string;
  public cnpj!: string | null;
  public phone!: string;
  public address!: string;
  public city!: string;
  public state!: string;
  public description!: string | null;
  public logoUrl!: string | null;
  public isActive!: boolean;
  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;
  public readonly deletedAt!: Date | null;
}

CompanyModel.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      unique: 'companies_user_id_unique',
      field: 'user_id',
      references: {
        model: 'users',
        key: 'id',
      },
    },
    tradeName: {
      type: DataTypes.STRING(255),
      allowNull: false,
      field: 'trade_name',
    },
    legalName: {
      type: DataTypes.STRING(255),
      allowNull: false,
      field: 'legal_name',
    },
    cnpj: {
      type: DataTypes.STRING(14),
      allowNull: true,
      unique: 'companies_cnpj_unique',
    },
    phone: {
      type: DataTypes.STRING(20),
      allowNull: false,
    },
    address: {
      type: DataTypes.STRING(500),
      allowNull: false,
    },
    city: {
      type: DataTypes.STRING(100),
      allowNull: false,
    },
    state: {
      type: DataTypes.STRING(2),
      allowNull: false,
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    logoUrl: {
      type: DataTypes.STRING(500),
      allowNull: true,
      field: 'logo_url',
    },
    isActive: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: true,
      field: 'is_active',
    },
    createdAt: {
      type: DataTypes.DATE,
      allowNull: false,
      field: 'created_at',
    },
    updatedAt: {
      type: DataTypes.DATE,
      allowNull: false,
      field: 'updated_at',
    },
    deletedAt: {
      type: DataTypes.DATE,
      allowNull: true,
      field: 'deleted_at',
    },
  },
  {
    sequelize,
    tableName: 'companies',
    timestamps: true,
    underscored: true,
    paranoid: true,
  },
);

export { CompanyModel };

File: mobile/lib/features/company/data/models/company_model.dart
import '../../domain/entities/company.dart';

class CompanyModel extends Company {
  const CompanyModel({
    required super.id,
    required super.userId,
    required super.tradeName,
    super.legalName,
    super.cnpj,
    required super.phone,
    required super.address,
    required super.city,
    required super.state,
    super.description,
    super.logoUrl,
    super.isActive,
    super.createdAt,
  });

  factory CompanyModel.fromJson(Map<String, dynamic> json) {
    return CompanyModel(
      id: json['id'] as String,
      userId: json['userId'] as String,
      tradeName: json['tradeName'] as String,
      legalName: json['legalName'] as String?,
      cnpj: json['cnpj'] as String?,
      phone: json['phone'] as String,
      address: json['address'] as String,
      city: json['city'] as String,
      state: json['state'] as String,
      description: json['description'] as String?,
      logoUrl: json['logoUrl'] as String?,
      isActive: json['isActive'] == true || json['isActive'] == 1,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'tradeName': tradeName,
      if (legalName != null) 'legalName': legalName,
      if (cnpj != null) 'cnpj': cnpj,
      'phone': phone,
      'address': address,
      'city': city,
      'state': state,
      if (description != null) 'description': description,
      if (logoUrl != null) 'logoUrl': logoUrl,
      'isActive': isActive,
      if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
    };
  }

  Map<String, dynamic> toUpdateJson() {
    return {
      'tradeName': tradeName,
      if (legalName != null && legalName!.isNotEmpty) 'legalName': legalName,
      if (cnpj != null && cnpj!.isNotEmpty) 'cnpj': cnpj,
      'phone': phone,
      'address': address,
      'city': city,
      'state': state,
      if (description != null && description!.isNotEmpty)
        'description': description,
    };
  }

  Map<String, dynamic> toRow() {
    return {
      'id': id,
      'userId': userId,
      'tradeName': tradeName,
      'legalName': legalName,
      'cnpj': cnpj,
      'phone': phone,
      'address': address,
      'city': city,
      'state': state,
      'description': description,
      'logoUrl': logoUrl,
      'isActive': isActive ? 1 : 0,
      'createdAt': createdAt?.toIso8601String(),
    };
  }

  factory CompanyModel.fromRow(Map<String, dynamic> row) {
    return CompanyModel(
      id: row['id'] as String,
      userId: row['userId'] as String,
      tradeName: row['tradeName'] as String,
      legalName: row['legalName'] as String?,
      cnpj: row['cnpj'] as String?,
      phone: row['phone'] as String,
      address: row['address'] as String,
      city: row['city'] as String,
      state: row['state'] as String,
      description: row['description'] as String?,
      logoUrl: row['logoUrl'] as String?,
      isActive: (row['isActive'] as int? ?? 1) == 1,
      createdAt: row['createdAt'] != null
          ? DateTime.tryParse(row['createdAt'] as String)
          : null,
    );
  }
}
